### PR TITLE
#216 Implement shunt regulation model

### LINKED ISSUE #216: Complete Shunt regulation implementation
When a shunt is not associated to a SMACC/ACMC in the database but is still with VoltageRegulatorOn=true in the iidm, the DYNModelCentralizedShuntsSectionControl cpp model should be used in the dyd for the regulation.

This regulation should be connected to the shunt (either directly to the Network model or to a Modelica model ShuntBWithSections).
 
There is an example in nrt/data/DynaFlow/CentralizedShuntsRegulation in dynawo (with Modelica model).

@marcochiaramello fyi

### PR DESCRIPTION
**Pull request by lecourtoisflo**
closes #216 

requires https://github.com/dynawo/dynawo/pull/1909

### CODE REVIEW

File: sources/Context.h [hunk 1]
```diff
@@ -170,7 +170,7 @@ class Context {
   algo::HVDCLineDefinitions hvdcLineDefinitions_;                        ///< hvdc definitions
   algo::GeneratorDefinitionAlgorithm::BusGenMap busesWithDynamicModel_;  ///< map of bus ids to a generator that regulates them
   algo::DynamicModelDefinitions dynamicModels_;                          ///< model definitions
-  algo::ShuntCounterDefinitions counters_;                               ///< shunt counters definitions
+  algo::ShuntDefinitions shuntDefinitions_;                              ///< shunt counters definitions
```
Comment (marcochiaramello) [inline sources/Context.h:173]: wrong comment?

File: sources/Inputs/include/Node.h [hunk 2]
```diff
@@ -141,15 +141,41 @@ class Tfo {
...
   using ShuntId = std::string;  ///< alias for shunt id
+  using BusId = std::string;    ///< Alias for bus id regulated by the shunt
 
   /**
-   * @brief Constructor
+   * @brief Construct a new Shunt
    *
    * @param id the shunt id
+   * @param busId the regulated bus of the shunt
+   * @param targetV the target V of the shunt
```
Comment (marcochiaramello) [inline sources/Inputs/include/Node.h:151]: the target volatge

File: sources/Inputs/include/Node.h [hunk 3]
```diff
@@ -141,15 +141,41 @@ class Tfo {
...
 struct Shunt {
   using ShuntId = std::string;  ///< alias for shunt id
+  using BusId = std::string;    ///< Alias for bus id regulated by the shunt
 
   /**
-   * @brief Constructor
+   * @brief Construct a new Shunt
    *
    * @param id the shunt id
+   * @param busId the regulated bus of the shunt
```
Comment (marcochiaramello) [inline sources/Inputs/include/Node.h:150]: the bus regulated by the shunt?

File: sources/Inputs/include/Node.h [hunk 4]
```diff
@@ -141,15 +141,41 @@ class Tfo {
...
+   * @param targetV the target V of the shunt
+   * @param voltageRegulationOn whether voltage regulation is enabled for the shunt
+   * @param bs the vector of the B values for the shunt
```
Comment (marcochiaramello) [inline sources/Inputs/include/Node.h:153]: replace b by susceptance

**Review by marcochiaramello — CHANGES_REQUESTED** [review]

File: sources/Context.h (hunk 1 above)
Comment (lecourtoisflo) [reply sources/Context.h:173]: OK

File: sources/Inputs/include/Node.h (hunk 3 above)
Comment (lecourtoisflo) [reply sources/Inputs/include/Node.h:150]: OK

File: sources/Inputs/include/Node.h (hunk 2 above)
Comment (lecourtoisflo) [reply sources/Inputs/include/Node.h:151]: OK

File: sources/Inputs/include/Node.h (hunk 4 above)
Comment (lecourtoisflo) [reply sources/Inputs/include/Node.h:153]: OK

File: sources/Inputs/include/Node.h [hunk 5]
```diff
@@ -141,15 +141,44 @@ class Tfo {
...
+   *
+   * @param other the shunt to compare to
+   * @return @b true if the shunts are equal, @b false if not
+   */
+  bool operator==(const Shunt& other) const {
+    return id == other.id;
+  }
+
+  const ShuntId id;                     ///< Shunt id
+  const BusId busId;                    ///< the connected bus of the shunt
```
Comment (marcochiaramello) [inline sources/Inputs/include/Node.h:177]: is this the bus of connection or the regulated one? the comment here is not aligned on the one of line 144 above

File: sources/Inputs/include/Node.h [hunk 6]
```diff
@@ -141,15 +141,44 @@ class Tfo {
...
+  const double targetV;                 ///< the target V of the shunt
+  const bool voltageRegulationOn;       ///< whether voltage regulation is enabled for the shunt
+  const std::vector<double> bSections;  ///< the vector of the B values for the shunt
```
Comment (marcochiaramello) [inline sources/Inputs/include/Node.h:180]: please replace B by susceptance everywhere in comments

**Review by marcochiaramello — CHANGES_REQUESTED** [review]

File: sources/Inputs/include/Node.h (hunk 5 above)
Comment (lecourtoisflo) [reply sources/Inputs/include/Node.h:177]: According to dynawo documentation, it is the connection bus (comment l144 corrected)

File: sources/Inputs/include/Node.h (hunk 6 above)
Comment (lecourtoisflo) [reply sources/Inputs/include/Node.h:180]: OK

**Pull request closed without merging**